### PR TITLE
Convert files to TypeScript

### PR DESCRIPTION
**Pull request by simon04**
https://www.typescriptlang.org/

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1144,7 +1144,7 @@ <h3 class="t subtitle" data-t="share.options"></h3>
       src="node_modules/jquery-ui/dist/jquery-ui.js"
       type="module"
     ></script>
-    <script src="./js/index.js" type="module"></script>
+    <script src="./js/index.ts" type="module"></script>
```
Comment (HolgerJeromin) [inline index.html:1147]: No. You do not serve a ts file to a browser

File: js/ffs.ts [hunk 2]
```diff
@@ -27,15 +28,12 @@ function normalize(query) {
...
-      const c = [];
-      for (let i = 0; i < c1.length; i++)
-        for (let j = 0; j < c2.length; j++) {
-          c.push({
-            logical: "and",
-            queries: c1[i].queries.concat(c2[j].queries)
-          });
-        }
-      return c;
+      return c1.flatMap((c1i) =>
```
Comment (HolgerJeromin) [inline js/ffs.ts:31]: This kills non chrome edge:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap#browser_compatibility

File: js/sync-with-osm.ts [hunk 3]
```diff
@@ -95,16 +95,16 @@ function loadQueries(callback) {
...
 
     const preferences = dom.querySelector("preferences");
     // clean up existing data
     const existing = preferences.querySelectorAll(
       `preference[k^="${configs.appname}_query"]`
     );
-    for (let i = 0; i < existing.length; i++) {
-      preferences.removeChild(existing[i]);
+    for (const e of existing) {
+      preferences.removeChild(e);
```
Comment (HolgerJeromin) [inline js/sync-with-osm.ts:107]:
```suggestion
      e.remove();
```

File: map.html [hunk 4]
```diff
@@ -14,6 +14,6 @@
     <div id="map"></div>
     <!--wait spinner-->
     <div class="modal"></div>
-    <script src="./js/map.js" type="module"></script>
+    <script src="./js/map.ts" type="module"></script>
```
Comment (HolgerJeromin) [inline map.html:17]: No ts for browser again

File: index.html (hunk 1 above)
Comment (HolgerJeromin) [reply index.html:1147]: I miss a build step to generate the js file.

Comment (simon04) [reply index.html:1147]: Vite automatically compiles TypeScript, see https://vitejs.dev/guide/features.html#typescript

#589

File: js/ffs.ts (hunk 2 above)
Comment (simon04) [reply js/ffs.ts:31]: 95% according to https://caniuse.com/mdn-javascript_builtins_array_flatmap

Comment (HolgerJeromin) [reply js/ffs.ts:30]: I know. 
But this should be an explicit decision.

File: index.html (hunk 1 above)
Comment (HolgerJeromin) [reply index.html:1147]: Thanks.

**Pull request merged**